File: internal/service/manager.go
package service

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/HyperVon/new-kraken-rebalancer/internal/config"
	"github.com/HyperVon/new-kraken-rebalancer/internal/model"
	"github.com/shopspring/decimal"
)

// PortfolioManager manages starting, stopping, and running the rebalancing loop.
type PortfolioManager interface {
	StartRebalancingLoop()
	StopRebalancingLoop()
	RunLoop(ctx context.Context)
	PerformRebalanceCycle() error
}

// PortfolioManagerImpl implements PortfolioManager.
type PortfolioManagerImpl struct {
	configService       config.ConfigService
	tradeHistoryService TradeHistoryService
	analyzer            *PortfolioAnalyzer
	executor            *OrderExecutor

	mu        sync.RWMutex
	isRunning bool
}

// NewPortfolioManagerImpl creates a new PortfolioManagerImpl.
func NewPortfolioManagerImpl(
	cfgService config.ConfigService,
	historyService TradeHistoryService,
	analyzer *PortfolioAnalyzer,
	executor *OrderExecutor,
) *PortfolioManagerImpl {
	return &PortfolioManagerImpl{
		configService:       cfgService,
		tradeHistoryService: historyService,
		analyzer:            analyzer,
		executor:            executor,
	}
}

func (m *PortfolioManagerImpl) StartRebalancingLoop() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.isRunning = true
	slog.Info("Rebalancing loop started")
}

func (m *PortfolioManagerImpl) StopRebalancingLoop() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.isRunning = false
	slog.Info("Rebalancing loop stopped")
}

func (m *PortfolioManagerImpl) RunLoop(ctx context.Context) {
	for {
		m.mu.RLock()
		running := m.isRunning
		m.mu.RUnlock()

		if !running {
			select {
			case <-ctx.Done():
				return
			case <-time.After(1 * time.Second):
				continue
			}
		}

		cfg := m.configService.GetConfig()
		slog.Info("Starting rebalance cycle", "dryRun", cfg.Settings.DryRun)

		if err := m.PerformRebalanceCycle(); err != nil {
			slog.Error("Error in rebalancing cycle", "error", err)
		}

		delaySeconds := cfg.Settings.LoopDelaySeconds
		if delaySeconds <= 0 {
			delaySeconds = 60
		}

		select {
		case <-ctx.Done():
			return
		case <-time.After(time.Duration(delaySeconds) * time.Second):
		}
	}
}

func (m *PortfolioManagerImpl) PerformRebalanceCycle() error {
	slog.Info("--- Starting Snapshot Phase ---")
	var actionLog []string

	balances, err := m.analyzer.FetchBalances()
	if err != nil {
		return fmt.Errorf("failed to fetch balances: %w", err)
	}

	prices, err := m.analyzer.FetchPrices()
	if err != nil {
		return fmt.Errorf("failed to fetch prices: %w", err)
	}

	portfolioValues := m.analyzer.CalculatePortfolioValues(balances, prices)
	if portfolioValues == nil {
		return errors.New("aborted rebalance cycle due to missing prices")
	}

	totalPortfolioValueUSD := portfolioValues.TotalValueUSD
	currentValuesUSD := portfolioValues.CurrentValuesUSD

	slog.Info("Portfolio valued", "totalUSD", totalPortfolioValueUSD.Round(2))

	drawdownPct := m.analyzer.UpdateAthAndCalculateDrawdown(totalPortfolioValueUSD)
	cfg := m.configService.GetConfig()
	fiatDeploymentPct := m.analyzer.CalculateFiatDeployment(drawdownPct, cfg.Settings)

	if fiatDeploymentPct.GreaterThan(decimal.Zero) {
		slog.Info("Drawdown detected",
			"drawdownPct", drawdownPct.Round(2),
			"fiatDeploymentPct", fiatDeploymentPct.Round(2),
		)
	}

	effectiveUsdTarget := m.analyzer.CalculateEffectiveUsdTarget(fiatDeploymentPct)
	cryptoScaleFactor := m.analyzer.CalculateCryptoScaleFactor(effectiveUsdTarget)

	analysisResult := m.analyzer.AnalyzeDeviations(totalPortfolioValueUSD, currentValuesUSD, effectiveUsdTarget, cryptoScaleFactor)
	actionLog = append(actionLog, analysisResult.ActionLog...)

	m.executor.ExecuteOrders(
		analysisResult.BuyOrders,
		analysisResult.SellOrders,
		currentValuesUSD,
		prices,
		cfg.Settings,
		&actionLog,
	)

	snapshot := m.buildSnapshot(
		balances,
		prices,
		currentValuesUSD,
		totalPortfolioValueUSD,
		effectiveUsdTarget,
		cryptoScaleFactor,
		drawdownPct,
		fiatDeploymentPct,
		actionLog,
	)

	if err := m.tradeHistoryService.AddSnapshot(snapshot); err != nil {
		slog.Error("Failed to persist trade history snapshot", "error", err)
	}

	slog.Info("--- Cycle Complete ---")
	return nil
}

func (m *PortfolioManagerImpl) buildSnapshot(
	balances RawBalances,
	prices RawPrices,
	currentValuesUSD map[string]decimal.Decimal,
	totalPortfolioValueUSD decimal.Decimal,
	effectiveUsdTarget decimal.Decimal,
	cryptoScaleFactor decimal.Decimal,
	drawdownPct decimal.Decimal,
	fiatDeploymentPct decimal.Decimal,
	actionLog []string,
) model.PortfolioSnapshot {
	assetSnapshots := make(map[string]model.AssetSnapshot)
	cfg := m.configService.GetConfig()

	for _, alloc := range cfg.Allocations {
		symbol := alloc.Symbol.String()
		balance := m.analyzer.ResolveBalance(symbol, balances)
		balDec := decimal.NewFromFloat(balance)
		valUSD := currentValuesUSD[symbol]

		priceDec := decimal.NewFromFloat(1.0)
		if !alloc.Symbol.IsUSD() {
			if p, exists := prices[symbol]; exists {
				priceDec = decimal.NewFromFloat(p)
			}
		}

		baseTargetPct := alloc.TargetPercent
		snapshotTargetPct := baseTargetPct
		var calcTargetPct decimal.Decimal

		if alloc.Symbol.IsUSD() {
			calcTargetPct = effectiveUsdTarget
		} else {
			calcTargetPct = baseTargetPct.Mul(cryptoScaleFactor)
			snapshotTargetPct = calcTargetPct
		}

		currentPct := decimal.Zero
		if totalPortfolioValueUSD.GreaterThan(decimal.Zero) {
			currentPct = valUSD.DivRound(totalPortfolioValueUSD, 4).Mul(decimal.NewFromInt(100))
		}

		targetVal := totalPortfolioValueUSD.Mul(calcTargetPct).DivRound(decimal.NewFromInt(100), 4)
		deviationUSD := valUSD.Sub(targetVal)
		devPct := decimal.Zero

		if targetVal.GreaterThan(decimal.Zero) {
			devPct = deviationUSD.DivRound(targetVal, 4).Mul(decimal.NewFromInt(100))
		}

		assetSnapshots[symbol] = model.AssetSnapshot{
			Symbol:           alloc.Symbol,
			Balance:          balDec,
			Price:            priceDec,
			ValueUSD:         valUSD,
			TargetPercent:    snapshotTargetPct,
			CurrentPercent:   currentPct,
			DeviationPercent: devPct,
			DeviationUSD:     deviationUSD,
		}
	}

	return model.PortfolioSnapshot{
		Timestamp:                 time.Now(),
		TotalValueUSD:             totalPortfolioValueUSD,
		Assets:                    assetSnapshots,
		Actions:                   actionLog,
		DrawdownPercent:           drawdownPct,
		FiatDeploymentPercent:     fiatDeploymentPct,
		EffectiveUsdTargetPercent: effectiveUsdTarget,
	}
}

File: backend/src/service/manager.ts
import { Decimal } from 'decimal.js';
import { ConfigService } from '../config/config';
import { TradeHistoryService } from './history';
import { PortfolioAnalyzer } from './analyzer';
import { OrderExecutor } from './executor';
import { PortfolioSnapshot, AssetSnapshot } from '../model/snapshot';
import { Asset } from '../model/asset';

Decimal.set({ rounding: Decimal.ROUND_HALF_UP });

export class PortfolioManager {
  private readonly configService: ConfigService;
  private readonly tradeHistoryService: TradeHistoryService;
  private readonly portfolioAnalyzer: PortfolioAnalyzer;
  private readonly orderExecutor: OrderExecutor;

  private isRunning = false;
  private timer: NodeJS.Timeout | null = null;

  constructor(
    configService: ConfigService,
    tradeHistoryService: TradeHistoryService,
    portfolioAnalyzer: PortfolioAnalyzer,
    orderExecutor: OrderExecutor
  ) {
    this.configService = configService;
    this.tradeHistoryService = tradeHistoryService;
    this.portfolioAnalyzer = portfolioAnalyzer;
    this.orderExecutor = orderExecutor;
  }

  startRebalancingLoop(): void {
    if (this.isRunning) return;
    this.isRunning = true;
    console.log('Rebalancing loop started.');
    this.runLoop();
  }

  stopRebalancingLoop(): void {
    this.isRunning = false;
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    console.log('Rebalancing loop stopped.');
  }

  private async runLoop(): Promise<void> {
    if (!this.isRunning) return;

    const settings = this.configService.getConfig().settings;
    try {
      console.log(`Starting Rebalance Cycle. DryRun: ${settings.dryRun}`);
      await this.performRebalanceCycle();
    } catch (e) {
      console.error('Error in rebalancing cycle', e);
    }

    if (this.isRunning) {
      const delayMs = settings.loopDelaySeconds * 1000;
      this.timer = setTimeout(() => this.runLoop(), delayMs);
    }
  }

  async performRebalanceCycle(): Promise<void> {
    console.log('--- Starting Snapshot Phase ---');
    const actionLog: string[] = [];

    const balances = await this.portfolioAnalyzer.fetchBalances();
    const prices = await this.portfolioAnalyzer.fetchPrices();
    const values = this.portfolioAnalyzer.calculatePortfolioValues(balances, prices);
    if (!values) {
      return;
    }

    const { totalValueUSD, currentValuesUSD } = values;
    console.log(`Total Portfolio Value: $${totalValueUSD.toFixed(2)}`);

    const drawdownPct = this.portfolioAnalyzer.updateAthAndCalculateDrawdown(totalValueUSD);
    const fiatDeploymentPct = this.portfolioAnalyzer.calculateFiatDeployment(
      drawdownPct,
      this.configService.getConfig().settings
    );

    if (fiatDeploymentPct.gt(0)) {
      console.log(
        `Drawdown Detected: ${drawdownPct.toFixed(2)}%. Fiat Deployment: ${fiatDeploymentPct.toFixed(2)}%`
      );
    }

    const effectiveUsdTarget = this.portfolioAnalyzer.calculateEffectiveUsdTarget(fiatDeploymentPct);
    const cryptoScaleFactor = this.portfolioAnalyzer.calculateCryptoScaleFactor(effectiveUsdTarget);

    const { buyOrders, sellOrders, actionLog: cycleActions } = this.portfolioAnalyzer.analyzeDeviations(
      totalValueUSD,
      currentValuesUSD,
      effectiveUsdTarget,
      cryptoScaleFactor
    );
    actionLog.push(...cycleActions);

    await this.orderExecutor.executeOrders(
      buyOrders,
      sellOrders,
      currentValuesUSD,
      prices,
      this.configService.getConfig().settings,
      actionLog
    );

    const snapshot = this.buildSnapshot(
      balances,
      prices,
      currentValuesUSD,
      totalValueUSD,
      effectiveUsdTarget,
      cryptoScaleFactor,
      drawdownPct,
      fiatDeploymentPct,
      actionLog
    );

    try {
      this.tradeHistoryService.addSnapshot(snapshot);
    } catch (e: unknown) {
      const message = e instanceof Error ? e.message : String(e);
      console.error('Failed to persist trade history snapshot', e);
      actionLog.push(`ERROR: Failed to persist trade history: ${message}`);
    }

    console.log('--- Cycle Complete ---');
  }

  private buildSnapshot(
    balances: Record<string, number>,
    prices: Record<string, Decimal>,
    currentValuesUSD: Record<string, Decimal>,
    totalPortfolioValueUSD: Decimal,
    effectiveUsdTarget: Decimal,
    cryptoScaleFactor: Decimal,
    drawdownPct: Decimal,
    fiatDeploymentPct: Decimal,
    actionLog: string[]
  ): PortfolioSnapshot {
    const assets: Record<string, AssetSnapshot> = {};

    for (const alloc of this.configService.getConfig().allocations) {
      const symbol = alloc.symbol;
      const isUsd = symbol.toUpperCase() === Asset.USD;

      const balanceVal = this.portfolioAnalyzer.resolveBalance(symbol, balances);
      const balance = new Decimal(balanceVal);
      const valUSD = currentValuesUSD[symbol] || new Decimal(0);
      const price = isUsd ? new Decimal(1) : prices[symbol] || new Decimal(1);

      const baseTargetPct = new Decimal(alloc.targetPercent);
      let snapshotTargetPct = baseTargetPct;
      let calcTargetPct: Decimal;

      if (isUsd) {
        calcTargetPct = effectiveUsdTarget;
      } else {
        calcTargetPct = baseTargetPct.mul(cryptoScaleFactor);
        snapshotTargetPct = calcTargetPct;
      }

      let currentPct = new Decimal(0);
      if (totalPortfolioValueUSD.gt(0)) {
        currentPct = valUSD.div(totalPortfolioValueUSD).toDecimalPlaces(4).mul(100);
      }

      const targetVal = totalPortfolioValueUSD.mul(calcTargetPct).div(100).toDecimalPlaces(4);
      const deviationUSD = valUSD.sub(targetVal);
      let devPct = new Decimal(0);

      if (targetVal.gt(0)) {
        devPct = deviationUSD.div(targetVal).toDecimalPlaces(4).mul(100);
      }

      assets[symbol] = {
        symbol,
        balance,
        price,
        valueUSD: valUSD,
        targetPercent: snapshotTargetPct,
        currentPercent: currentPct,
        deviationPercent: devPct,
        deviationUSD
      };
    }

    return {
      timestamp: new Date().toISOString(),
      totalValueUSD: totalPortfolioValueUSD,
      assets,
      actions: actionLog,
      drawdownPercent: drawdownPct,
      fiatDeploymentPercent: fiatDeploymentPct,
      effectiveUsdTargetPercent: effectiveUsdTarget
    };
  }
}
